### PR TITLE
Fix preempt cull (fake non-consum. 'mismatch')

### PR DESCRIPTION
**Pull request by alexis-cousein**
#### Describe Bug or Feature

The function to cull the list of preemption candidates does the wrong thing for non-consumable resources.

Suppose that we have a cluster with two nodes with 40 cpus.

node A has resources_available.platform=a, the other, node B, resources_available.platform=b.

We run a job 1 that asks for -lselect=1:ncpus=40 in a low priority queue. Since it doesn't ask for a specific platform at all, suppose it ends up on node A.

The second job 2 asks for -lselect=1:ncpus=40:platform=a, in an express queue.

If we did not cull the preemption candidate list, then 1 would be stripped in simulation and 2 would run.

With the culling in place in the current code, 2 won't run, since 1 will be discarded as a candidate because inp->err is INSUFFICIENT_RESOURCE and the rdef is for the "platform" resource, nowhere to be found in the chunks of 1, which incorrectly makes the culling function decide that 1 "is not useful". It is, but because of the nodes it is running on (the node utility check is done in select_index_to_preempt, but only for preemption candidates that were not culled earlier).

#### Describe Your Change

For a job that does not run because of INSUFFICIENT_RESOURCE on a non-consumable resource, do not cull early, and let the select_index_to_preempt routine do the node utility check to see if the nodes of the candidate are useful for chunks of the preemptor.

#### Attach Test and Valgrind Logs/Output

PTL test not written since developer machine in unsuitable to run PTL tests. Demonstration on a small (separate) test cluster included -- this could be used to make a PTL test.

Code changes backported to 19.2.6 and tested separately on this cluster, and they confirmed that the bug is gone.

```
# qmgr -c "p s"
--
#
# Create resources and set their properties.
#
#
# Create and define resource application
#
create resource application
set resource application type = string
set resource application flag = h
#
# Create and define resource platform
#
create resource platform
set resource platform type = string_array
set resource platform flag = h
#
# Create queues and set their attributes.
#
#
# Create and define queue workq
#
create queue workq
set queue workq queue_type = Execution
set queue workq Priority = 100
set queue workq enabled = True
set queue workq started = True
#
# Create and define queue expressq1
#
create queue expressq1
set queue expressq1 queue_type = Execution
set queue expressq1 Priority = 150
set queue expressq1 enabled = True
set queue expressq1 started = True
#
# Create and define queue expressq2
#
create queue expressq2
set queue expressq2 queue_type = Execution
set queue expressq2 Priority = 150
set queue expressq2 enabled = True
set queue expressq2 started = True
#
# Set server attributes.
#
set server scheduling = True
set server default_queue = workq
set server log_events = 511
set server mail_from = adm
set server query_other_jobs = True
set server resources_default.ncpus = 1
set server default_chunk.ncpus = 1
set server scheduler_iteration = 600
set server resv_enable = True
set server node_fail_requeue = 310
set server max_array_size = 10000
set server pbs_license_info = /opt/altair_1.lic
set server eligible_time_enable = False
set server max_concurrent_provision = 5
set server max_job_sequence_id = 9999999
 
# pbsnodes -av
pbshost1
Mom = pbshost1
Port = 15002
pbs_version = 19.2.6.20200311140837
ntype = PBS
state = free
pcpus = 4
resources_available.arch = linux
resources_available.host = pbshost1
resources_available.mem = 0b
resources_available.ncpus = 0
resources_available.vnode = pbshost1
resources_assigned.accelerator_memory = 0kb
resources_assigned.hbmem = 0kb
resources_assigned.mem = 0kb
resources_assigned.naccelerators = 0
resources_assigned.ncpus = 0
resources_assigned.vmem = 0kb
resv_enable = True
sharing = default_shared
in_multivnode_host = 1
license = l
last_state_change_time = Thu Apr 16 00:25:53 2020
 
pbshost1[0]
Mom = pbshost1
Port = 15002
pbs_version = 19.2.6.20200311140837
ntype = PBS
state = free
pcpus = 2
resources_available.application = app1
resources_available.arch = linux
resources_available.host = pbshost1
resources_available.mem = 2gb
resources_available.ncpus = 2
resources_available.platform = a,universal
resources_available.vnode = pbshost1[0]
resources_assigned.accelerator_memory = 0kb
resources_assigned.hbmem = 0kb
resources_assigned.mem = 0kb
resources_assigned.naccelerators = 0
resources_assigned.ncpus = 0
resources_assigned.vmem = 0kb
resv_enable = True
sharing = default_shared
in_multivnode_host = 1
license = l
last_state_change_time = Thu Apr 16 04:23:26 2020
last_used_time = Thu Apr 16 04:23:26 2020
 
pbshost1[1]
Mom = pbshost1
Port = 15002
pbs_version = 19.2.6.20200311140837
ntype = PBS
state = free
pcpus = 2
resources_available.application = app2
resources_available.arch = linux
resources_available.host = pbshost1
resources_available.mem = 2gb
resources_available.ncpus = 2
resources_available.platform = b,universal
resources_available.vnode = pbshost1[1]
resources_assigned.accelerator_memory = 0kb
resources_assigned.hbmem = 0kb
resources_assigned.mem = 0kb
resources_assigned.naccelerators = 0
resources_assigned.ncpus = 0
resources_assigned.vmem = 0kb
resv_enable = True
sharing = default_shared
in_multivnode_host = 1
license = l
last_state_change_time = Thu Apr 16 04:11:29 2020
last_used_time = Thu Apr 16 04:11:29 2020
 
# su - test1
Last login: Thu Apr 16 04:21:30 EDT 2020 on pts/1
[test1@pbshost1 ~]$ qsub -l select=1:ncpus=2:application=app1 -q workq -- /bin/sleep 2000
79.pbshost1
 
[test1@pbshost1 ~]$ qstat -aw
pbshost1:
Req'd  Req'd   Elap
Job ID                         Username        Queue           Jobname         SessID   NDS  TSK   Memory Time  S Time
------------------------------ --------------- --------------- --------------- -------- ---- ----- ------ ----- - -----
79.pbshost1                    test1           workq           STDIN              44128    1     2    --    --  R 00:00:00
 
[test1@pbshost1 ~]$ qsub -l select=1:ncpus=2:application=app1:platform=a -q expressq1 -- /bin/sleep 2000
80.pbshost1
 
[test1@pbshost1 ~]$ sleep 20; qstat -aw
pbshost1:
Req'd  Req'd   Elap
Job ID          Username Queue    Jobname    SessID NDS TSK Memory Time  S Time
--------------- -------- -------- ---------- ------ --- --- ------ ----- - -----
79.pbshost1     test1    workq    STDIN       44128   1   2    --    --  R 00:00 pbshost1/0*2
Job run at Thu Apr 16 at 04:24 on (pbshost1[0]:ncpus=2)
80.pbshost1     test1    expressq STDIN         --    1   2    --    --  Q   --   --
Not Running: Insufficient amount of resource: platform
```

### CODE REVIEW

File: src/scheduler/job_info.c [hunk 1]
```diff
@@ -5255,6 +5255,16 @@ static int cull_preemptible_jobs(resource_resv *job, void *arg)
...
 					return 1;
 			} else {
+				if (inp->err->rdef->type.is_non_consumable) {
+				    /* non consumables are used for node selection so
+				     * comparing preemptor and candidate is not sufficient to cull;
+				     * nodes the candidate is using might have the correct tags 
+				     * even if the candidate did not request the non-consumable.
+				     * This will be checked in the node utility test in
+				     * select_index_to_preempt
+				     */
```
Comment (bhroam) [inline src/scheduler/job_info.c:5265]: This comment is a little too wordy for my tastes.  How about we say "In the non-consumable case, we need to pass the job on.  There is a case when a job requesting a non-specific select is allocated a node with a non-consumable.  We will check nodes to see if they are useful in select_index_to_preempt()"

Comment (alexis-cousein) [reply src/scheduler/job_info.c:5265]: Will update.

Comment (alexis-cousein) [reply src/scheduler/job_info.c]: Updated.

**Comment by bhroam** [comment]
@alexis-cousein you do not need a small test cluster to test this.  You should be able to write a PTL test using self.server.create_vnodes().  The scheduler doesn't care if vnodes are real hosts or not.  It just cares about vnodes.  Please write a test.

**Comment by alexis-cousein** [comment]
OK, I'll try to figure out how to reconfigure queues/vnodes/resources in the PTL test for a few hours...

**Comment by alexis-cousein** [comment]
[pbs_pull_1668_test.py.zip](https://github.com/PBSPro/pbspro/files/4582188/pbs_pull_1668_test.py.zip)

Here's a short PTL test. Please note that I have two "tries" around commands to set the preempt_order to "R" -- if integrating this test on master, you need only retain the test that sets the scheduler attribute (rather than the older one for 19.2.x that sets it in the sched config file).

**Comment by alexis-cousein** [comment]
> @alexis-cousein you do not need a small test cluster to test this. You should be able to write a PTL test using self.server.create_vnodes(). The scheduler doesn't care if vnodes are real hosts or not. It just cares about vnodes. Please write a test.

[pbs_pull_1668_test.py.zip](https://github.com/PBSPro/pbspro/files/4582251/pbs_pull_1668_test.py.zip)

Here's a small test. There are two "try" constructs around the attempts to change the preempt_order, since you need different commands on the 19.2.5 scheduler I tested and master (which now uses a scheduler attribute rather than a line in the sched_config file). If integrating it in master you can simplify by simply removing the first try: block and removing the try:...except:... construct around the setting of the preempt_order attribute.

On my server I verified that a fixed 19.2.5 binary succeeds in the test while the stock 19.2.5 fails.

**Comment by alexis-cousein** [comment]
Would you like this to be part of the pull request (in which case I will indeed simplify the test so that it is merely compatible with master)? In that case, please tell me how I need to make sure that the test in the .py file gets run if inserted in the "functional" directory. Is there a list of tests to be run somewhere?

**Comment by bhroam** [comment]
Yes, please add the automated test to your PR.

**Comment by alexis-cousein** [comment]
In what directory? in the directory above "functional"? In a sibling directory that I create next to functional (and if so, how do I name it)? in the "functional" directory itself?

**Comment by bhroam** [comment]
Put it in functional.  Not necessary, but if possible put it in an existing test plan which is broad enough to cover it.  If not, name it like the other tests (start it with pbs_).  The class name is camel cased starting with Test.  Unit test requires each test case to start with 'test_'.  There is a preemption test plan, so that might be a good choice.

**Comment by alexis-cousein** [comment]
OK. I'll insert it in the preemption test plan tomorrow, with the syntax to cater for 19.2.x ripped out.

**Comment by alexis-cousein** [comment]
Or rather, later today, from my vantage point ;-).

**Comment by alexis-cousein** [comment]
PTL test "test_preempt_wrong_cull" added to pbs_preemption.py.

File: test/tests/functional/pbs_preemption.py [hunk 2]
```diff
@@ -733,3 +733,57 @@ def test_preempt_requeue_resc(self):
...
+        Test to make sure that if a preemptor cannot run because
+        it misses a non-consumable on a node, preemption candidates
+        are not incorrectly removed from consideration
+        if and because they "do not request the relevant resource".
+        Deciding on their utility should be left to the check
+        to see whether the nodes they occupy are useful.
+        """
+        attr = {}
+        attr['type'] = 'string_array'
+        attr['flag'] = 'h'
```
Comment (bhroam) [inline test/tests/functional/pbs_preemption.py:748]: Consider putting this all on one line.

File: test/tests/functional/pbs_preemption.py [hunk 3]
```diff
@@ -733,3 +733,57 @@ def test_preempt_requeue_resc(self):
...
+             'enabled': 'True',
+             'Priority': 1}
+        self.server.manager(MGR_CMD_CREATE, QUEUE, a, "lopri")
+
+        # submit job 1
+        a = {'Resource_List.select': '1:ncpus=1:vnode=vna[0]', ATTR_q: 'lopri'}
+        j1 = Job(TEST_USER, attrs=a)
+        jid1 = self.server.submit(j1)
+
+        self.server.expect(JOB, {'substate': '42'}, id=jid1)
```
Comment (bhroam) [inline test/tests/functional/pbs_preemption.py:782]: Don't worry about the substate, look for state: R.  PTL handles the substate for you. (here and below)

File: test/tests/functional/pbs_preemption.py (hunk 2 above)
Comment (alexis-cousein) [reply test/tests/functional/pbs_preemption.py]: Quite:

        attr = {'type': 'string_array', 'flag': 'h'}

File: test/tests/functional/pbs_preemption.py (hunk 3 above)
Comment (alexis-cousein) [reply test/tests/functional/pbs_preemption.py]: Yes, indeed, I see it from the source (it adds substate 42 if you do not specify a substate explicitly). 

Changed to 'state': 'R'.
